sql: use uint32 for DOid

rafiss opened the pull request:
refs https://github.com/cockroachdb/cockroach/issues/41904

This resolves a piece of tech debt that has caused us a few surprises
and bugs in the past. This doesn't change anything about the on-disk
representation -- it just makes it so that OIDs handled in memory are
more reliably unsigned 32 bit integers.

Release note: None

rafiss committed:
sql: use uint32 for DOid

This resolves a piece of tech debt that has caused us a few surprises
and bugs in the past. This doesn't change anything about the on-disk
representation -- it just makes it so that OIDs handled in memory are
more reliably unsigned 32 bit integers.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82430)
<!-- Reviewable:end -->

rafiss committed:
eval: show error when casting to OID out of range

Release note (sql change): Casting from an integer to OID, or calling
the `oid` builtin function, using an integer that is larger than 32 bits
now results in an error. Specifically, the range for valid inputs is
[MinInt32, MaxUint32].

rafiss committed:
eval: show error when comparing OID to int out of range

Release note (sql change): An error is now returned if an OID is
compared to an integer that is out of the allowable range of OID input
values, which is [MinInt32, MaxUint32].
